subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -590,48 +659,142 @@ _set_au_boundaries (GstRtpH266Pay * rtph266pay)
 static GstFlowReturn
 _push_pending_data (GstRtpH266Pay * rtph266pay, gboolean eos)
 {
+  AggregatedNalus *aggregated_nalus = &rtph266pay->aggregated_nalus;
   GQueue *nalus = &rtph266pay->nalus;
   GstFlowReturn ret = GST_FLOW_OK;
   Nalu *nalu;
 
-  // TODO: handle AP
-
   // FIXME: Maybe it's worth to avoid iterating the NALUs several times, maybe not
   _insert_xps_cache (rtph266pay);
   _set_au_boundaries (rtph266pay);
 
   // Try to push all NALUs
   while ((nalu = g_queue_pop_head (nalus))) {
-    gboolean is_last = g_queue_get_length (nalus) == 0;
-
-    // Can't push the last NALU without knowing if it's the end of an AU,
-    // because setting the M bit could be necessary. But have to push it if
-    // we're on EOS
-    if (!eos && is_last && !nalu->au_end) {
-      GST_DEBUG_OBJECT (rtph266pay, "Keeping last NALU: %" NALU_PTR_FORMAT,
+    if (_can_aggregate_nalu (rtph266pay, nalu)) {
+      _aggregated_nalus_add (aggregated_nalus, nalu);
+      GST_DEBUG_OBJECT (rtph266pay, "NALU aggregated: %" NALU_PTR_FORMAT,
           NALU_ARGS (nalu));
-      g_queue_push_head (nalus, nalu);
-      break;
+      continue;
     }
 
-    GST_DEBUG_OBJECT (rtph266pay, "Pushing NALU: %" NALU_PTR_FORMAT,
-        NALU_ARGS (nalu));
+    // vvv Can't aggregate anymore vvv
+    if (!_aggregated_nalus_is_empty (aggregated_nalus)) {
+      g_queue_push_head (nalus, nalu);  // Keep this one for later
+      ret = _push_aggregated (rtph266pay);      // Push what we have right now
```

Comment:
It doesn't matter whether we have a complete AU or not; as we can no longer aggregate, we need to push the AP we have right now.

Reply:
Let's say an AU has Nalu1, Nalu2... Nalu100. Can an AP have Nalu1..50 and the next AP have Nalu51... Nalu100?

Reply:
Yes! Exactly! An AP can have **less than one** AU, but it can not have **more than one** AU.

> An AP aggregates NAL units of one access unit, and it MUST NOT contain NAL units from more than one AU.

Reply:
Can you comment that?

Reply:
It was already commented: https://github.com/fluendo/gstreamer/pull/42/commits/a1ee00bee8eaef3349e976c5c8f8a6d11da84e7e#diff-a7ceab36a5207c8c558b7431ebdef49454d2a74ad1c606fa6540963de7e81646R721